cmd/node_problem_detector.go

```diff
@@ -46,23 +47,27 @@ func main() {
 
 	npdo.ValidOrDie()
 
-	monitors := make(map[string]types.Monitor)
-	for _, config := range npdo.SystemLogMonitorConfigPaths {
-		if _, ok := monitors[config]; ok {
-			// Skip the config if it's duplicated.
-			glog.Warningf("Duplicated monitor configuration %q", config)
-			continue
-		}
-		monitors[config] = systemlogmonitor.NewLogMonitorOrDie(config)
+	// Initialize problem daemons.
+	problemDaemons, errors := problemdaemon.NewProblemDaemons(npdo.MonitorsConfigs)
+	for _, err := range errors {
+		glog.Errorf("Problem daemon initialization failed: %v", err)
+	}
+	if len(errors) != 0 && npdo.PanicOnProblemDaemonFailure {
+		glog.Fatalf("Quit NPD because of above problem daemon initialization failures")
 	}
 
+	// Legacy per-monitor configuration.
+	// This section is kept so that existing command line option still works on NPD.
+	// TODO(xueweiz): when NPD's refactoring is finished, remove legacy configuration.
+	for _, config := range npdo.SystemLogMonitorConfigPaths {
```

Comment:
Why not just change original flags to write to new monitor config?
So that we can preserve original flags, but all use the new monitor config internally.

Reply:
Actually I can't. The flag that's defined at last will always overwrite the previously defined flag.

For example:

```
        for _, problemDaemonName := range problemdaemon.GetProblemDaemonNames() {
		npdo.MonitorsConfigs[problemDaemonName] = &[]string{}
		fs.StringSliceVar(
			npdo.MonitorsConfigs[problemDaemonName],
			"config."+string(problemDaemonName),
			[]string{},
			fmt.Sprintf("Comma separated configurations for %v monitor. %v",
				problemDaemonName,
				problemdaemon.GetProblemDaemonHandler(problemDaemonName).CmdOptionDescription))
	}
	fs.StringSliceVar(npdo.MonitorsConfigs["system-log-monitor"], "system-log-monitors",
		[]string{}, "List of paths to system log monitor config files, comma separated. This option will be replaced by --config.system-log-monitor.")
	fs.StringSliceVar(npdo.MonitorsConfigs["system-log-monitor"], "custom-plugin-monitors",
		[]string{}, "List of paths to custom plugin monitor config files, comma separated. This option will be replaced by --config.custom-plugin-monitor.")
```

In this code, "--config.system-log-monitor" argument will always be ignored. Because whatever the value it has, "--system-log-monitors" will always overwrite it.

I think it's OK if we leave this code for now. I plan to first change all the test to use the new argument, then remove this code. It's shouldn't be too bad.

Reply:
> In this code, "--config.system-log-monitor" argument will always be ignored. Because whatever the value it has, "--system-log-monitors" will always overwrite it.

I don't understand this part. We can update the code to change the behavior.

My preferred behavior is: panic if both `--config.system-log-monitor` and `--system-log-monitors` are specified. We should only allow one is used. And then here, we can achieve what @Random-Liu described: change original flags to write to the new monitor config.

Reply:
Good point, thanks! I like the panic behavior.
Change the code so that when both old and new flags are used, NPD will panic. And added unit test for it.

Reply:
I think @Random-Liu's original comment on "Why not just change original flags to write to new monitor config?" is still valid.

if `--system-log-monitor` is specified, set it to `--config.system-log-monitor`. Then here, we can just have one line:
```
problemDaemons := problemdaemon.NewProblemDaemons(npdo.MonitorConfigPaths)
```
The legacy code of creating problem daemons from old flags can be removed.

Reply:
Ah good point. Followed your setConfigsOrDie() idea. Done.